### PR TITLE
[WT-384] Basic Card Reader Integration

### PR DESCRIPTION
**Pull request by l-c-mcc**
### Description
This PR adds card reader integration. From the admin checkout page, purchases may be directed to a specified card reader for payment. Successful payment is informed via webhook. While waiting for payment, the user has access to a dashboard that allows for polling the Stripe API in case of webhook failure and an option to cancel the order, if the card has not yet been presented to the reader.

Please follow the new instructions added to the README if testing this branch. Additionally, this feature requires a new variable (websocket_url) in the .env file; see the .env.dist file. To test readers without physical hardware, simulated readers must be set up in either the Stripe dashboard, the Stripe CLI, or other means, [see here](https://stripe.com/docs/terminal/payments/connect-reader?reader-type=simulated). To test reader payment, [see here](https://stripe.com/docs/terminal/references/testing#simulated-card-presentment).

### Risks
There are a few areas of concern:
- This was tested using only simulated hardware. Unexpected issues may come up with real hardware.
- There are a lot of potential failure points with card readers. We tried to design this system to be resilient to them, but there may be blind spots.
- Right now, the new environment variable for the websocket url fails to work in the client side despite working on the server side. A workaround is provided on the client side in the meantime, allowing this branch to function in local environments. Additionally, we are unsure if more work has to be done to get this new variable set up on the cloud server. **This is the main reason why this PR is currently a draft. Advice regarding this would be highly appreciated.**

### Validation
Manual testing.

### Issue
https://github.com/WonderTix/WonderTix/issues/384

### Operating System
Windows 10.

### CODE REVIEW

File: README.md
Comment (ben16williams) [inline README.md:46]: It would be nice if this step briefly described what the command is and why you need to do it

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (ben16williams) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx:74]: These `console.log()`s are fine for debugging locally but should be removed before this is merged and deployed

File: server/src/controllers/eventController.service.ts
Comment (ben16williams) [inline server/src/controllers/eventController.service.ts:74]: The comments could be removed along with the commented out field if it's not being used

File: server/src/controllers/eventController.ts
Comment (ben16williams) [inline server/src/controllers/eventController.ts:192]: This should be the actual contacts `contactid` and not hard-coded

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (ben16williams) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx:74]: *the console logs in the server are fine but the one's in the client are user facing

**Comment by ben16williams** [comment]
The shape of this looks pretty good. Really neat to see it coming together. Didn't get a chance to dig into the functionality yet but will once the few other comments are addressed. Would be nice if @henrykaus or @benfaustin95 could take a look if y'all have a chance since you're more familiar with the client admin checkout pages

File: README.md
Comment (l-c-mcc) [reply README.md]: Got it; added a couple sentences of explanation.

File: server/src/controllers/eventController.ts
Comment (l-c-mcc) [reply server/src/controllers/eventController.ts:192]: Reid and I were under the impression that the card reader would be used to collect payment immediately before an event, so collecting contact information would be too time consuming if there is a line, hence why we just set the contact to database value for anonymous.

If this is not the case, we can have the reader checkout go to the contact info form before the reader payment processing page. Or, perhaps, whoever is operating the reader could have a checkbox that determines whether contact info is collected or not as part of the payment process so that the reader is more context flexible. Or, maybe, just an email is collected as a middle-ground, since, based on a quick look, email is how people are attached to accounts. What do you think?

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (ReidLuttrell) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx]: Removed console.logs in client

File: server/src/controllers/eventController.service.ts
Comment (ReidLuttrell) [reply server/src/controllers/eventController.service.ts]: Removed unecessary comments

File: server/src/controllers/eventController.ts
Comment (ben16williams) [reply server/src/controllers/eventController.ts:203]: The email address is a unique identifier of a contact. Does stripe return an email address associated with the card from the reader checkout? If there is really no way to get an email other then manually entering it, then we can skip associating the order/tickets with a contact altogether

Comment (ReidLuttrell) [reply server/src/controllers/eventController.ts:203]: My understanding is that there is no additional information automatically associated with a card on a card_present to a terminal reader, at least none that stripe gives you access to. If we were to not have a contactid, would that mean passing undefined to contactid in orderfulfillment?

Comment (ben16williams) [reply server/src/controllers/eventController.ts:203]: If the result would be that the contactid is null in the order then I think yes

Comment (ReidLuttrell) [reply server/src/controllers/eventController.ts]: contactid is now null in terminal reader orders. Note: I had to change and migrate the database scheme to let contactid_fk be nullable. Here is what a completed reader order looks like now
Code_msXbT7p2Ho

Comment (ReidLuttrell) [reply server/src/controllers/eventController.ts]: @ben16williams for some reason the github actions is compiling with an error thats just nonexistant on my machine. On top of that, i didn't even change anything that affected this part of the code. have you ever encountered this? 
msedge_AOs09jtS18

Comment (ReidLuttrell) [reply server/src/controllers/eventController.ts]: nevermind i figured it out

**Comment by l-c-mcc** [comment]
The latest push has the client side environmental variable working. Deployed environmental variables still not configured.

**Comment by l-c-mcc** [comment]
Made an attempt at the deployment environmental variables based on [this PR](https://github.com/WonderTix/WonderTix/pull/528).

Also included the Playwright fix from [PR #618 ](https://github.com/WonderTix/WonderTix/pull/618).

This should be enough to convert from a draft to a full PR.

File: client/cloudbuild.yaml
Comment (ben16williams) [inline client/cloudbuild.yaml:78]: It doesn't look like this has been added in[ secret manager](https://console.cloud.google.com/security/secret-manager?referrer=search&project=wondertix-app) yet. Do you know what values we would want to use for dev, stg, and prod?

File: README.md
Comment (ben16williams) [inline README.md:53]: Having not looked into it, is there a way to complete a reader checkout locally (eg. provide a payment_method to stripe)? Follow up, do you know if there would there be a way to test a complete reader checkout in a deployed environment? Would you mind adding instructions for doing either/or here?

File: server/prisma/migrations/20240215202648_init/migration.sql
Comment (ben16williams) [inline server/prisma/migrations/20240215202648_init/migration.sql:410]: This has to be in a new, distinct database migration file

File: client/cloudbuild.yaml
Comment (l-c-mcc) [reply client/cloudbuild.yaml:78]: The general format of the websocket URL should look like:

`wss://{root_location}/wss`

So if the root URL is `https://rooturl.com`, the Websocket URL should be `wss://rooturl.com/wss`. This should apply for the dev, stg, and prod.

It is worth noting that the `/wss` at the end is not strictly necessary, it can be anything or nothing. It just seemed to be a reasonable name, and keeping it consistent allows us to easily use the [ws library feature](https://github.com/websockets/ws/blob/HEAD/doc/ws.md#new-websocketserveroptions-callback) to limit connections to a specific path in the future if we need to for whatever reason.

File: README.md
Comment (l-c-mcc) [reply README.md:53]: > Having not looked into it, is there a way to complete a reader checkout locally (eg. provide a payment_method to stripe)? Follow up, do you know if there would there be a way to test a complete reader checkout in a deployed environment? Would you mind adding instructions for doing either/or here?

Reader checkouts can be completed locally. I assume the same method for completing them locally would work in a deployed environment. The instructions are as follows:

1. Ensure you have the Stripe CLI installed.
2. Turn on the servers and ensure that the webhook is listening to Stripe events. These events are diff
3. First, set up the simulated reader. It is possible to this via the Stripe CLI or through the Stripe API, but these instructions will explain how to do so through the Stripe dashboard as I guess that is how readers will often be set up in practice.
a. Go to https://dashboard.stripe.com/test/terminal. 
b. Add a location if one does not already exist.
c. Click on the location in the location list.
d. Press register reader.
e. Use `simulated-wpe` as the registration code.
f. Once created, click on the reader in the list to get an info pop up to appear. Write down the Reader ID somewhere, which should have the format `tmr_xxx`.
4. Navigate to the deployed front end (ex: `https://localhost:3000`). Log in, and go to the admin ticketing page.
5. Create your sample purchase. Once it is ready, press the "send to reader" button. The adjacent drop down menu should display a list of possible readers to send to; as you have only set up one reader, the only option should match the reader ID noted earlier.
6. You should be taken to the reader checkout page.
a. It is worth noting that here you can cancel the payment early *if a payment method has not yet been presented*. It also can directly poll the Stripe API for cases when the webhook is slow to confirm a payment, or if the webhook is not listening to events for whatever reason.
7. Open a terminal.
8. Run this command: `stripe test_helpers terminal readers present_payment_method tmr_xxx`, replacing `tmr_xxx` with your Reader ID.
9. The frontend page should soon redirect you to the success page.

These instructions were based on the premise you are testing with a simulated reader and a virtual card. Based on what I read, they should not vary much if you were instead working with a physical reader and a physical test card. For more information on how to set up either simulated readers or real readers, [see here](https://docs.stripe.com/terminal/payments/connect-reader?reader-type=simulated). For more information on simulated card presenting, [see here](https://docs.stripe.com/terminal/references/testing?lang=cli#simulated-card-presentment). For physical test cards, [see here](https://docs.stripe.com/terminal/references/testing?lang=cli#physical-test-cards).

I will add a simplified form of these instructions to the readme shortly.

File: server/prisma/migrations/20240215202648_init/migration.sql
Comment (l-c-mcc) [reply server/prisma/migrations/20240215202648_init/migration.sql]: I created the new migration and moved this code out of the init migration.

File: README.md
Comment (ReidLuttrell) [reply README.md:53]: Also, if you want to test the error popups for declined payments, you can use one of [these](https://docs.stripe.com/testing) test numbers that simulate failures, for example to simulate an insufficient funds error, you would run `stripe test_helpers terminal readers present_payment_method tmr_xxx --card_present.number 4000000000009995`

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx:670]: There should only be one `h1` per page otherwise it messes a little bit with screen readers, this could probably just be an `h2`

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx:57]: What is _supposed_ to happen here?

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx:89]: A token should be supplied to this request as well

File: server/src/controllers/orderController.service.ts
Comment (henrykaus) [inline server/src/controllers/orderController.service.ts:72]: We typically don't need console.logs in the server

File: server/src/controllers/orderController.service.ts
Comment (henrykaus) [inline server/src/controllers/orderController.service.ts:269]: Same here

File: server/src/server.ts
Comment (henrykaus) [inline server/src/server.ts:693]: here as well

File: server/src/server.ts
Comment (henrykaus) [inline server/src/server.ts:700]: This should be a `console.error`

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx:210]: I think these make sense for debugging purposes, but for practical use, I think we wouldn't need the `paymentIntentID` displayed. And then regarding the status, we could make this a bit more human-friendly by mapping the possible values of status to less code-looking values. Ex. `payment_intent.requires_payment_method` --> `Requires Payment Method`

File: README.md
Comment (henrykaus) [inline README.md:53]: @benfaustin95 Do you know if changing this command if going to affect order processing?

File: README.md
Comment (henrykaus) [reply README.md:53]: I got these steps to work except for the `stripe test_helpers ...` command. I had to access to API endpoint directly, not sure if stripe was configured with the sk or not
```
curl -X POST https://api.stripe.com/v1/test_helpers/terminal/readers/tmr_.../present_payment_method \
  -u "sk_test_...:"
```

File: client/src/App.tsx
Comment (ReidLuttrell) [reply client/src/App.tsx]: Done

File: server/src/controllers/eventController.ts
Comment (ReidLuttrell) [reply server/src/controllers/eventController.ts]: All APIs should now use token

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx
Comment (ReidLuttrell) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx]: Token now supplied to all requests

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx
Comment (ReidLuttrell) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx]: reader select was made label

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (ReidLuttrell) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx]: This was a misunderstanding I had earlier that I had meant to remove, our payment intent object actually has automatic confirmation enabled, and confirmCardPayment only applies to web purchases because it redirects you to a confirmation page, this case has been removed.

File: server/src/controllers/orderController.service.ts
Comment (ReidLuttrell) [reply server/src/controllers/orderController.service.ts]: removed all console logs from server

File: server/src/server.ts
Comment (ReidLuttrell) [reply server/src/server.ts]: Changed to console.error

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (ReidLuttrell) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx]: I decided to replace the payment Intent ID here with the order number, since its less visually busy and still contains potentially useful information, statuses are also now "translated" to normal looking text

File: README.md
Comment (ReidLuttrell) [reply README.md:53]: if you can only use curl, then to supply a card number i believe you would do `curl -X POST https://api.stripe.com/v1/test_helpers/terminal/readers/tmr_xxx/present_payment_method -u "sk_test_xxx:" -d "type=card_present" -d "card_present[number]=4000000000009995"`

File: README.md
Comment (ReidLuttrell) [reply README.md:53]: The `--events` parameter is somewhat of a filter for what events pass through to the webhook, for the card reader it is required that payment_intent and terminal.reader events are passed to the webhook, which is the reason this parameter was removed. Since checkout sessions carry `__ticketing` metadata, and those are the only events passed through the ticketing webhook, and based on testing, I believe this lack of filtering should not effect the order processing.

Comment (benfaustin95) [reply README.md:53]: Looks good to me

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx
Comment (henrykaus) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/ReaderPurchase.tsx]: Sounds good, thanks!

File: client/cloudbuild.yaml
Comment (henrykaus) [reply client/cloudbuild.yaml:78]: @ben16williams Once we have a secret in Secret Manager, this looks good to me

Comment (ben16williams) [reply client/cloudbuild.yaml:78]: Done https://console.cloud.google.com/security/secret-manager/secret/WEBSOCKET_URL/versions?project=wondertix-app. dev is configured as `wss://dev.wondertix.com/wss`, staging is `wss://stg.wondertix.com/wss`, and production is `wss://www.wondertix.com/wss`

**Review by ben16williams — APPROVED** [review]
It looks like all of @henrykaus concerns have been addressed so LGTM, thanks for doing this! It's an important piece of wondertix becoming a practicality

**Comment by l-c-mcc** [comment]
> It looks like all of @henrykaus concerns have been addressed so LGTM, thanks for doing this! It's an important piece of wondertix becoming a practicality

@ben16williams hey, Reid and I tested the dev.wondertix.com reader checkout and the websocket does not seem to be working. Our current hypothesis is that the deployment env var, at least for the client, has not been set up in the deploy-client.sh file (or something along those lines), as it seems an empty string is being used as the websocket url on the frontend. I'll be reviewing the deployment process in this commit soon to look for errors, but I would appreciate if you or someone else familiar with the process could also look it over at some point as I am still shaky in regard to the cloud deployment.

**Comment by ben16williams** [comment]
> 1. I missed this on my previous reviews of client env vars, but it seems that a change may also be required in the build-client.sh file. `"WEBSOCKET_URL"` must be added to the required list at the start of the fille, and `--build-arg REACT_APP_WEBSOCKET_URL=${WEBOSCKET-URL}` added to the end. If the deployment process requires a PR to be opened and merged to test this, I could quickly make one.

Very well could be the case. If you create a PR with that change I'll approve it and we can push it through

> 2. I may have miscommunicated my explanation of what the secrets need to be set to, and erred when checking the secret values you used. By root url, I was referring to what the .env files call the root url, that being the backend/server url, like where the Swagger page could be found.

Ah I see, I'll update those today

**Pull request merged**
